mkosi/vmspawn.py

```diff
@@ -0,0 +1,227 @@
+        else:
+            fname = config.output_dir_or_cwd() / config.output
+
+        # Make sure qemu can access the ephemeral copy. Not required for directory output because we don't pass that
+        # directly to qemu, but indirectly via virtiofsd.
+        os.chown(fname, INVOKING_USER.uid, INVOKING_USER.gid)
+
+        if config.output_format == OutputFormat.disk and config.runtime_size:
+            run(["systemd-repart",
+                 "--definitions", "",
+                 "--no-pager",
+                 f"--size={config.runtime_size}",
+                 "--pretty=no",
+                 "--offline=yes",
+                 fname])
+
+        root = None
+        if kernel:
+            qemu_cmdline += ["-kernel", kernel]
+
+            # We can't rely on gpt-auto-generator when direct kernel booting so synthesize a root=
+            # kernel argument instead.
+            root = finalize_root(find_partitions(fname))
+            if not root:
+                die("Cannot perform a direct kernel boot without a root or usr partition")
+
+        cmdline += [f"--image={fname}"]
+
+        if kernel and (KernelType.identify(kernel) != KernelType.uki or not config.architecture.supports_smbios()):
+            kcl = config.kernel_command_line + config.kernel_command_line_extra
+        else:
+            kcl = config.kernel_command_line_extra
+
+        if root:
+            kcl += [root]
+
+        for src, target in config.runtime_trees:
+            sock = stack.enter_context(start_virtiofsd(src, uidmap=True))
+            qemu_cmdline += [
+                "-chardev", f"socket,id={sock.name},path={sock}",
+                "-device", f"vhost-user-fs-pci,queue-size=1024,chardev={sock.name},tag={sock.name}",
+            ]
+            kcl += [f"systemd.mount-extra={sock.name}:{target or f'/root/src/{src.name}'}:virtiofs"]
+
+        if kernel and (KernelType.identify(kernel) != KernelType.uki or not config.architecture.supports_smbios()):
+            qemu_cmdline += ["-append", " ".join(kcl)]
+        elif config.architecture.supports_smbios():
+            qemu_cmdline += [
+                "-smbios",
+                f"type=11,value=io.systemd.stub.kernel-cmdline-extra={' '.join(kcl)}"
+            ]
+
+        if (
+            kernel and KernelType.identify(kernel) != KernelType.uki and
+            "-initrd" not in args.cmdline and
+            (config.output_dir_or_cwd() / config.output_split_initrd).exists()
+        ):
+            qemu_cmdline += ["-initrd", config.output_dir_or_cwd() / config.output_split_initrd]
+
+
+        if (
+            firmware == QemuFirmware.uefi and
+            config.qemu_swtpm != ConfigFeature.disabled and
+            shutil.which("swtpm") is not None
+        ):
+            sock = stack.enter_context(start_swtpm())
+            qemu_cmdline += ["-chardev", f"socket,id=chrtpm,path={sock}",
+                        "-tpmdev", "emulator,id=tpm0,chardev=chrtpm"]
```

Comment:
Formatting?

Reply:
This block is no longer present in 594d4fb.